apps/emqx_node_rebalance/src/emqx_node_rebalance_purge.erl

```diff
@@ -0,0 +1,233 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2022-2023 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+-module(emqx_node_rebalance_purge).
+
+-include("emqx_node_rebalance.hrl").
+
+-include_lib("emqx/include/logger.hrl").
+-include_lib("emqx/include/types.hrl").
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+-export([
+    start/1,
+    status/0,
+    stop/0
+]).
+
+-export([start_link/0]).
+
+-behaviour(gen_statem).
+
+-export([
+    init/1,
+    callback_mode/0,
+    handle_event/4,
+    code_change/4
+]).
+
+-export_type([
+    start_opts/0,
+    start_error/0,
+    stop_error/0
+]).
+
+%%--------------------------------------------------------------------
+%% APIs
+%%--------------------------------------------------------------------
+
+-define(DEFAULT_PURGE_RATE, 500).
+-define(ENABLE_KIND, purge).
+
+%% gen_statem states
+-define(disabled, disabled).
+-define(purging, purging).
+-define(cleaning_retained_messages, cleaning_retained_messages).
```

Comment:
retained message cleaning does not belong to cluster rebalance.

Reply:
This under the `purge` scenario of the `emqx_node_rebalance` application.

Reply:
Also, we're using the coordinator machinery from rebalance to make it more consistent, please see this [discussion](https://github.com/emqx/emqx/pull/11447#issuecomment-1681841362).